Libraries/rocSPARSE/level_2/gebsrmv/main.cpp

```diff
@@ -0,0 +1,194 @@
+    constexpr size_t row_ptr_size = sizeof(*d_bsr_row_ptr) * (mb + 1);
+    constexpr size_t col_ind_size = sizeof(*d_bsr_col_ind) * nnzb;
+
+    HIP_CHECK(hipMalloc((void**)&d_bsr_row_ptr, row_ptr_size));
+    HIP_CHECK(hipMalloc((void**)&d_bsr_col_ind, col_ind_size));
+    HIP_CHECK(hipMalloc((void**)&d_bsr_val, val_size));
+    HIP_CHECK(hipMalloc((void**)&d_x, x_size));
+    HIP_CHECK(hipMalloc((void**)&d_y, y_size));
+
+    HIP_CHECK(hipMemcpy(d_bsr_row_ptr, h_bsr_row_ptr.data(), row_ptr_size, hipMemcpyHostToDevice));
+    HIP_CHECK(hipMemcpy(d_bsr_col_ind, h_bsr_col_ind.data(), col_ind_size, hipMemcpyHostToDevice));
+    HIP_CHECK(hipMemcpy(d_bsr_val, h_bsr_val.data(), val_size, hipMemcpyHostToDevice));
+    HIP_CHECK(hipMemcpy(d_x, h_x.data(), x_size, hipMemcpyHostToDevice));
+    HIP_CHECK(hipMemcpy(d_y, h_y.data(), y_size, hipMemcpyHostToDevice));
+
+    // 3. Initialize rocSPARSE by creating a handle.
+    rocsparse_handle handle;
+    ROCSPARSE_CHECK(rocsparse_create_handle(&handle));
+
+    // 4. Prepare utility variables for rocSPARSE gebsrmv invocation.
+    // Matrix descriptor.
+    rocsparse_mat_descr descr;
+    ROCSPARSE_CHECK(rocsparse_create_mat_descr(&descr));
+
+    // 5. Call gebsrmv to perform y = alpha * A * x + beta * y.
+    ROCSPARSE_CHECK(rocsparse_dgebsrmv(handle,
+                                       dir,
+                                       trans,
+                                       mb,
+                                       nb,
+                                       nnzb,
+                                       &alpha,
+                                       descr,
+                                       d_bsr_val,
+                                       d_bsr_row_ptr,
+                                       d_bsr_col_ind,
+                                       bsr_row_dim,
+                                       bsr_col_dim,
+                                       d_x,
+                                       &beta,
+                                       d_y));
+
+    // 6. Copy solution to host from device.
+    HIP_CHECK(hipMemcpy(h_y.data(), d_y, y_size, hipMemcpyDeviceToHost));
+
+    // 7. Clear rocSPARSE allocations on device.
+    ROCSPARSE_CHECK(rocsparse_destroy_handle(handle));
+    ROCSPARSE_CHECK(rocsparse_destroy_mat_descr(descr));
+
+    // Clear device arrays.
+    HIP_CHECK(hipFree(d_bsr_row_ptr));
+    HIP_CHECK(hipFree(d_bsr_col_ind));
+    HIP_CHECK(hipFree(d_bsr_val));
+    HIP_CHECK(hipFree(d_x));
+    HIP_CHECK(hipFree(d_y));
+
+    // 8. Print results to standard output.
+    std::cout << "Solution successfully computed: ";
+
+    std::cout << "y = " << format_range(std::begin(h_y), std::end(h_y)) << std::endl;
+
+    // Compare solution with the expected result.
+    int          errors{};
+    const double eps = 1.0e5 * std::numeric_limits<double>::epsilon();
+    for(size_t i = 0; i < h_y.size(); ++i)
```

Comment:
Examples are not consistent with providing an error check with the expected result. Ellmv is different from this, for instance.

Reply:
Moreover, results should be exact here since inputs are integer values. The eps sounds random, and the relative error should be computed rather than absolute.

Reply:
For the first remark, the error-checking has been homogenized across examples in the last commit 👍🏼 . For the second one, while it's true that in this case the values of the matrix are integers, the type used for the matrix' values is `double`, so I think we should still showcase the proper error checking for double precision floats. Perhaps the input matrix/scalars could be changed so that floating point values are actually used.

Reply:
According to the inputs, you can set the tolerance to zero here; but that's a detail.
It is good that error-checking has been homogenized across examples.